server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java

```diff
@@ -91,6 +87,9 @@ public final void collectBucket(LeafBucketCollector subCollector, int doc, long
      * Same as {@link #collectBucket(LeafBucketCollector, int, long)}, but doesn't check if the docCounts needs to be re-sized.
      */
     public final void collectExistingBucket(LeafBucketCollector subCollector, int doc, long bucketOrd) throws IOException {
+        if (doc == 1) {
```

Comment:
Hm, I think this isn't quite right.  `doc` is the doc ID, so that could be all over the place, and also all going into the same bucket.  I think there are two options here:

1. Down below, we do `if (docCounts.increment(bucketOrd, 1) == 1) { <breaker stuff> }` which I think will work because the increment method returns the count after incrementing.  So if we have a doc count of 1, it's the first doc and a new bucket so we can account it

2. Alternatively, we could just account for it up in `collectBucket` without a conditional, since theoretically that should only be called on new buckets.  It's not guaranteed by the API but in practice that's how aggs use it.

There are two other issues we need to address though:

1. The old breaker logic only checked every 1024 buckets, since checking the real-memory breaker has a certain amount of overhead.  So we should re-implement that somehow
2. Trickier situation which I didn't think about when suggesting `BucketsAggregator`... if we add the 1024 threshold back, it's only a local count so aggs with 1023 buckets will never trigger the breaker even if the overall query has millions of buckets.  

Perhaps we continue to use the MultiBucketConsumer service thing, but move the breaker accounting to a different method?  That way it could maintain the global count and BucketsAggregator just calls a method on it or something?  Not sure, we can discuss more offline